azkaban-common/src/main/java/azkaban/version/JdbcVersionSetLoader.java

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright 2020 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.version;
+
+import azkaban.db.DatabaseOperator;
+import azkaban.db.SQLTransaction;
+import java.io.IOException;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.commons.dbutils.ResultSetHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A JDBC based implementation for loading VersionSet from the backed db. This class keeps the copy
+ * of entries in the version_set table locally as well to reduce the number of queries to the db.
+ * <p>
+ * This is a singleton class to be instantiated using dependency injection.
+ */
+@Singleton
+public class JdbcVersionSetLoader implements VersionSetLoader {
```

Comment:
One Suggestion: I think the cache part can be improved. Currently the cache and the database access part is fully coupled together. Cache can be build on top of database as a wrapper and allow caller to invoke the cache methods directly. If the requested versionSet is not present in the cache access from database and populate the cache. In case of insert first insert into cache and then database. This should not cause any issue as direct db access is not allowed as it will be always via cache.

Reply:
I agree on the decoupling part which will allow us to have any implementation of the cache in the future as well. However, I am not sure about the requirement of the cache itself. This is similar to schedule based trigger where we keep all the schedules in the memory and always update both the local map and the db through the same class. Since the code is only accessed by one class, we can refactor it easily in the future if the cache requirement arises.